powermon/device.py

```diff
@@ -96,10 +96,13 @@ def runLoop(self, force=False) -> bool:
         runs them
         """
         time.sleep(0.1)
-        if self.commands is None:
+        if self.commands is None or len(self.commands) == 0:
             log.info("no commands in queue")
             return False
         else:
+            # open connection on port
+            self.port.connect()
```

Comment:
In debug mode I see that sometimes port open/closed without needs:
`2023-09-16 18:42:22,866:DEBUG:serialport:disconnect@51: usbserial port disconnecting
2023-09-16 18:42:22,966:DEBUG:serialport:connect@42: usbserial port connecting. path:/dev/moxa0, baud:2400
2023-09-16 18:42:22,966:DEBUG:serialport:disconnect@51: usbserial port disconnecting
2023-09-16 18:42:23,066:DEBUG:serialport:connect@42: usbserial port connecting. path:/dev/moxa0, baud:2400
2023-09-16 18:42:23,067:DEBUG:serialport:disconnect@51: usbserial port disconnecting
2023-09-16 18:42:23,167:DEBUG:serialport:connect@42: usbserial port connecting. path:/dev/moxa0, baud:2400
2023-09-16 18:42:23,167:DEBUG:serialport:disconnect@51: usbserial port disconnecting
2023-09-16 18:42:23,267:DEBUG:serialport:connect@42: usbserial port connecting. path:/dev/moxa0, baud:2400
2023-09-16 18:42:23,267:DEBUG:serialport:disconnect@51: usbserial port disconnecting
2023-09-16 18:42:23,368:DEBUG:serialport:connect@42: usbserial port connecting. path:/dev/moxa0, baud:2400
2023-09-16 18:42:23,368:DEBUG:device:runLoop@108: Running command: QPIGS`

I thought this is due to empty commands list but nope. It might be because `force or command.dueToRun()` is false. Would be great to open port only there are command that will be executed it will improve performance, as well as open port once to run all commands. 
Any thoughts on this?

Reply:
The list of commands won't be empty when there is nothing to run. It will still have all the commands it's just that dueToRun() will return false.

I would put `self.port.connect()` in the `if force or command.dueToRun():` statement so it's only called if something is about to be run. This will cause it to be called for every command. To fix that, in the port.connect() add a check to see if it's already connected and do nothing if the connection is open.

Reply:
Another option would be to add an a method to the AbstractPort called is_connected() that returns true if the port is already connected and check to see if it's already connected before trying. This is a bit more verbose but might make the main loop more readable. We also can also use it to enforce that each port implementation has that method so the code works as expected when we add more ports

Reply:
What do you think if put connect and disconnect into initialize and finalize methods instead of checking if it connected or not in each command run?

Reply:
That would leave a persistent connection which may not work for all ports. Calling it every command and managing the connection behaviour in the port implementation leaves more options open. @jblance would have a better idea about the needs of different ports. I only use USB, going to start using bluetooth as well in the next few months.

Reply:
Bluetooth is particular bad with multiple reconnects - it causes the bluetooth stack to crash eventually (and on the JKBMS it beeps for every reconnect)
Victron VEDirect sends data continuously on serial as well
mpp-solar did the connect on each command and is fine for the simple case but doesnt work for many situations, hence the move to a maintained connection

Reply:
I think it should be something like
- disconnect in finalize
- a connect for each command run, but with a is_connected test in the port.connect and so the connect runs if not connected but doent run if connected, the is_connected logic can then cover a bunch of cases
- possibly a `reconnect for each command` type config parameter that triggers a disconnect after the command run

I have a bad network at the inverter/battery shed and have been thinking about the need to deal with connection failures (not relevant here as much as for the mqtt connection, but similar logic may need to apply)

Reply:
What logic should be in `is_connected` method? Is it enough check if `serialPort` variable is not None? How can we check if connection is still alive?